File: src/types/traditionArtifacts/SpellSwordEnchantment.ts
import { Entity, IncludeIdentifier, Object, Optional, Required } from "tsondb/schema/def"
import {
  ap_value,
  cost,
  effect,
  levels,
  maximum,
  name,
  name_in_library,
  property,
  select_options,
  volume,
} from "../_Activatable.js"
import { GeneralPrerequisites } from "../_Prerequisite.js"
import { NestedLocaleMap } from "../Locale.js"
import { Errata } from "../source/_Erratum.js"
import { src } from "../source/_PublicationRef.js"

export const SpellSwordEnchantment = Entity(import.meta.url, {
  name: "SpellSwordEnchantment",
  namePlural: "SpellSwordEnchantments",
  type: () =>
    Object({
      levels,
      select_options,
      maximum,
      prerequisites: Optional({
        type: IncludeIdentifier(GeneralPrerequisites),
      }),
      volume,
      cost,
      property,
      ap_value,
      src,
      translation: NestedLocaleMap(
        Required,
        "SpellSwordEnchantmentTranslation",
        Object({
          name,
          name_in_library,
          effect,
          errata: Optional({
            type: IncludeIdentifier(Errata),
          }),
        })
      ),
    }),
})

File: Sources/OptolithDatabaseSchema/Types/traditionArtifacts/SpellSwordEnchantment.swift
import FileDB

@Model
public struct SpellSwordEnchantment {
      levels,
      select_options,
      maximum,
      prerequisites: Optional({
        type: IncludeIdentifier(GeneralPrerequisites),
      }),
      volume,
      cost,
      property,
      ap_value,

    /// The publications where you can find the entry.
    let src: PublicationRefs

    /// All translations for the entry, identified by IETF language tag (BCP47).
    @Relationship
    let translations: [String: Translation]

    struct Translation { // SpellSwordEnchantmentTranslation
          name,
          name_in_library,
          effect,

        let errata: Errata?
    }
}
